[rocprof-compute] added f4f6 description and VALU FLOPS split for empirical peaks

jamessiddeley-amd opened the pull request:
## Motivation

Split VALU FLOPS into 3 separate metrics, each with their own datatype: F16FLOPS, F32FLOPS, F64FLOPS so that empirical peak values can be displayed alongside them. Before, VALU FLOPS was being calculated and there was no corresponding empirical peak value from roofline.csv so I made a fallback to use the theoretical equation. After this PR, all Roofline rate metrics will be able to be displayed alongside their empirical peak values.

Also, the metric description for F4F6 has been added.

As well, all the empirical peak variables are explicitly defined as local variables for them to be properly captured by the `ammolite_vars = {key: val for key, val in locals().items() if key.startswith("ammolite__")}` line.

## Technical Details

Changes to unified_config,yaml and small tweak in parser.py.

## Test Plan

local tests pass

## Test Result

currently running

## Submission Checklist

- [X] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

jamessiddeley-amd committed:
added f4f6 description and VALU FLOPS split

jamessiddeley-amd committed:
changed peak ammolite vars to local

vedithal-amd commented on projects/rocprofiler-compute/src/utils/parser.py:
```diff
@@ -925,32 +925,70 @@ def eval_metric(dfs, dfs_type, sys_info, empirical_peaks_df, raw_pmc_df, debug,
+    ammolite__L2Bw_empirical_peak = 0
+    ammolite__L1Bw_empirical_peak = 0
+    ammolite__LDSBw_empirical_peak = 0
+    ammolite__MFMA_FLOPs_F6F4_empirical_peak = 0
```
Instead of handling variables individually, let do it dynamically by leveraging `peak_names` list!

```
if empirical_peaks_df:
    peak_names = [<insert list elements here>]
    for peak_name in default_peaks:
        globals()[f"ammolite__{peak_name}"] = emprical_peaks_df.iloc[0][peak_name] if peak_name in empirical_peaks_df.columns else 0
```

vedithal-amd commented on projects/rocprofiler-compute/src/utils/parser.py:
```diff
@@ -925,32 +925,70 @@ def eval_metric(dfs, dfs_type, sys_info, empirical_peaks_df, raw_pmc_df, debug,
+        if "LDSBw" in empirical_peaks_df.columns:
+            ammolite__LDSBw_empirical_peak = peak_data_row["LDSBw"]
+        if "MFMA_FLOPs_F6F4" in empirical_peaks_df.columns:
+            ammolite__MFMA_FLOPs_F6F4_empirical_peak = peak_data_row["MFMA_FLOPs_F6F4"]
```
This is not necessary, if we handle variables dynamically, see above comment.

vedithal-amd commented on projects/rocprofiler-compute/src/utils/parser.py:
```diff
@@ -925,32 +925,70 @@ def eval_metric(dfs, dfs_type, sys_info, empirical_peaks_df, raw_pmc_df, debug,
             "value using --specs-correction"
         )
+
+    default_peaks = [
```
peak_names would be better naming i think

vedithal-amd commented:
Scratch that, using `ammolite__` variables or using `globals()` is a bad practice, please let's not continue this practice.

We could just setup a dictionary of empirical peaks

```
empirical_peak_values = {} if empirical_peaks_df.empty else emprical_peaks_df.iloc[0].to_dict()
```

Then pass this dictionary as `locals` to any `eval` calls within the `eval_metric` function

```
eval(compile(expression, "<string", "eval"), {}, empirical_peak_values))
```

In case you are wondering, the second argument which is empty dict is the `globals`

This should be done for all `ammolite__` vars at some point of time, buts let's start now

jamessiddeley-amd commented:
Creating a dict without the `ammolite__` prefix requires changes to `build_eval_string` where instead of converting `$var` to `ammolite__var`, keep it as `var`. This also means we must remove the ammolite definitions for the `sysinfo.csv` variables as well (e.g. `ammolite__cur_sclk`). So we would need to refactor the entire variable system now, not just the empirical peaks and do the rest at a later time.
I can simply, for now, init the peak dict with the `ammolite__` prefix and it will work with the current implementation. Or we can fully remove `ammolite__` variables now. What do you think @vedithal-amd ?

jamessiddeley-amd committed:
Merge branch 'develop' of https://github.com/ROCm/rocm-systems into users/jamessiddeley-amd/roofline-config

jamessiddeley-amd committed:
reverted to dict peak initialization

jamessiddeley-amd committed:
ruff check format

jamessiddeley-amd committed:
updated VALU descriptions

jamessiddeley-amd committed:
updated VALU descriptions

jamessiddeley-amd committed:
Merge branch 'develop' into users/jamessiddeley-amd/roofline-config

feizheng10 commented:
> Creating a dict without the `ammolite__` prefix requires changes to `build_eval_string` where instead of converting `$var` to `ammolite__var`, keep it as `var`. This also means we must remove the ammolite definitions for the `sysinfo.csv` variables as well (e.g. `ammolite__cur_sclk`). So we would need to refactor the entire variable system now, not just the empirical peaks and do the rest at a later time. I can simply, for now, init the peak dict with the `ammolite__` prefix and it will work with the current implementation. Or we can fully remove `ammolite__` variables now. What do you think @vedithal-amd ?



> Creating a dict without the `ammolite__` prefix requires changes to `build_eval_string` where instead of converting `$var` to `ammolite__var`, keep it as `var`. This also means we must remove the ammolite definitions for the `sysinfo.csv` variables as well (e.g. `ammolite__cur_sclk`). So we would need to refactor the entire variable system now, not just the empirical peaks and do the rest at a later time. I can simply, for now, init the peak dict with the `ammolite__` prefix and it will work with the current implementation. Or we can fully remove `ammolite__` variables now. What do you think @vedithal-amd ?

Separate PR seems good to me

feizheng10 commented on projects/rocprofiler-compute/src/rocprof_compute_soc/analysis_configs/gfx908/0400_roofline.yaml:
```diff
@@ -28,6 +36,11 @@ Panel Config:
       per second. Note: this does not include any 64-bit floating point operations
       from VALU instructions. The peak empirically measured F64 MFMA operations achievable
       on the specific accelerator is displayed alongside for comparison.'
+    MFMA FLOPs (F6F4): 'The total number of 4-bit and 6-bit floating point MFMA operations
```
Technically, it should not show up in any arch other than gfx950

jamessiddeley-amd replied:
I don't believe in the structure of `unified_config.yaml` that we can limit a metric description to a single arch. @vedithal-amd can you confirm?

vedithal-amd replied:
You can target the metric per arch, if the metric is not targeted for a given arch the metric description wont show up.

We dont have support for per arch metric description right now

jamessiddeley-amd replied:
Yes the metrics description is global for id block 400, so it will be present however there is no metric `F6F4` in any gfx other than 950 so it won't show.

vedithal-amd reviewed (dismissed):
LGTM, thanks for making the change

cfallows-amd reviewed (dismissed):
Thanks for improvement initiative and checking in on roofline CLI block.

jamessiddeley-amd committed:
Update parser.py

jamessiddeley-amd committed:
Merge branch 'develop' into users/jamessiddeley-amd/roofline-config

jamessiddeley-amd commented:
Experienced PDSB failures and local ctest failures before merging on double-check. Deduced it had to do with the empirical peaks dict, in its current form it wasnt in the scope of multiprocessor's eval. Updating it to `globals()` allows it to be viewed and it is now getting 100% passes in local test suite.

Transition from `globals()` and `ammolite__` vars entirely may come in a seperate PR.

<img width="610" height="386" alt="image" src="https://github.com/user-attachments/assets/08cb940a-b900-4a0e-88e1-a0d0832eb9f7" />

cfallows-amd reviewed (dismissed):
Re-approve after fix for ctest- passing now.

jamessiddeley-amd committed:
Merge branch 'develop' into users/jamessiddeley-amd/roofline-config

jamessiddeley-amd committed:
Merge branch 'develop' into users/jamessiddeley-amd/roofline-config

jamessiddeley-amd committed:
Update parser.py

Added gracefull NameError handling
Moved globals() update to init_metric_evaluation with ammolite__ vars and raw pmc_df

jamessiddeley-amd committed:
update formatting

jamessiddeley-amd commented:
Edited globals() update to be in `init_metric_evaluator` function alongside `ammolite__` and `pmc_df` global updates rather than the main thread:

```
def init_metric_evaluator(
    raw_pmc_df: Union[pd.DataFrame, dict], ammolite_vars: dict, empirical_peaks: dict
) -> None:
    if isinstance(raw_pmc_df, dict):
        raw_pmc_df_keys = set(raw_pmc_df.keys())
. . .
. . .
    globals().update(raw_pmc_df_items)
    globals().update(ammolite_vars)
    globals().update(empirical_peaks)
```
Also, spoke to Joseph Macaranas about failing PDSB test, which has been failing for the previous 8 builds. He mentioned the failed PDSB test can be ignored, since they're redesigning the test infrastructure to use "test sharding"
